from_module_test.go: test fails when git repo exists on path with symlinks

### Terraform Version

This is an issue with one of the `terraform` unit tests, so I'll list the `git` commit of `HEAD` on the `master` branch.

```
    $ GIT_PAGER=cat git branch --no-color
    * master

    $ git describe --abbrev=12 --dirty --broken
    v0.13.1-4-g2bd2a9a9238f

    $ git rev-parse HEAD
    2bd2a9a9238fc59eb60689ff06956590b68b5a63
```

### Terraform Configuration Files

(n/a)


### Debug Output

(see below)


### Crash Output

(n/a)


### Expected Behavior


Unit tests should have passed when building.


### Actual Behavior

The test `from_module_test.go` failed.


### Steps to Reproduce

The test fails when the 'terraform' git repo exists in a directory path that contains a symlink.

```
    $ mkdir -p realdir-parent/realdir-child

    $ ln -s realdir-parent/realdir-child my-repos

    $ cd my-repos

    $ git clone 'git@github.com:hashicorp/terraform.git'

    $ cd terraform

    $ GOPATH=/tmp/whatever go test github.com/hashicorp/terraform/internal/initwd
    --- FAIL: TestDirFromModule_submodules (0.00s)
        from_module_test.go:125: tempChdir switched to /tmp/terraform-configload954601254 after copying from testdata/empty
        from_module_test.go:150: slice[0].LocalPath: /path/to/realdir-parent/realdir-child/terraform/internal/initwd/testdata/local-modules/child_a != /path/to/my-repos/terraform/internal/initwd/testdata/local-modules/child_a
        from_module_test.go:150: slice[1].LocalPath: /path/to/realdir-parent/realdir-child/terraform/internal/initwd/testdata/local-modules/child_a/child_b != /path/to/my-repos/terraform/internal/initwd/testdata/local-modules/child_a/child_b
    FAIL
    FAIL    github.com/hashicorp/terraform/internal/initwd  0.012s
    FAIL
```


### Additional Context

The function `TestDirFromModule_submodules(...)` in the file:
```
    internal/initwd/from_module_test.go
```
instantiates the local variable `fromModuleDir` as an absolute file path (via `filepath.Abs(...)`).

That path is passed as the `sourceAddr` param to `DirFromModule(...)`, and also used to construct the unit test's "wanted" outputs.

The call path through `DirFromModule(...)` eventually invokes `installLocalModule(...)`, which uses the "canonicalized" directory path returned from `filepath.EvalSymlinks(...)`.

Therefore, the unit test comparison values (in particular, the 'LocalPath' members of the `wantCalls` elements) must also be constructed from canonicalized path values.

The issue would only show up when the source tree is located on a path that contains a symlink. When the source tree location is comprised of only real directories, the values returned by `filepath.Abs(...)` and `filepath.EvalSymlinks(...)` would be identical, so would not trigger the issue.

I have a patch; I'll open a PR for this shortly.


### References

#### Are there any other GitHub issues (open or closed) or Pull Requests that should be linked here?

Nothing definitive that I could find.

It looks like there was work in this area in (#20539, #20603) and (#21060, #21063), but those are more about the general capaiblity of following symlinks for local module paths. The current issue is specifically about just the build-time situation with a failing unit test.

--- a/internal/initwd/from_module_test.go
+++ b/internal/initwd/from_module_test.go
@@ -122,6 +122,17 @@ func TestDirFromModule_submodules(t *testing.T) {
 		t.Fatal(err)
 	}
 
+	// DirFromModule will expand ("canonicalize") the pathnames, so we must do
+	// the same for our "wantCalls" comparison values. Otherwise this test
+	// will fail when building in a source tree with symlinks in $PWD.
+	//
+	// See also: https://github.com/hashicorp/terraform/issues/26014
+	//
+	fromModuleDirRealpath, err := filepath.EvalSymlinks(fromModuleDir)
+	if err != nil {
+		t.Error(err)
+	}
+
 	tmpDir, done := tempChdir(t, fixtureDir)
 	defer done()
 
@@ -138,12 +149,12 @@ func TestDirFromModule_submodules(t *testing.T) {
 		{
 			Name:       "Install",
 			ModuleAddr: "child_a",
-			LocalPath:  filepath.Join(fromModuleDir, "child_a"),
+			LocalPath:  filepath.Join(fromModuleDirRealpath, "child_a"),
 		},
 		{
 			Name:       "Install",
 			ModuleAddr: "child_a.child_b",
-			LocalPath:  filepath.Join(fromModuleDir, "child_a/child_b"),
+			LocalPath:  filepath.Join(fromModuleDirRealpath, "child_a/child_b"),
 		},
 	}
 

